### ISSUE TITLE
Regular method missing return type should not be assumed to be a static constructor

### ISSUE THREAD

**Issue by danmoseley** [dotnet/roslyn#36121]
**Version Used**: 
 04252e from master, 21 May 19

**Steps to Reproduce**:
```c#
class X
{
    private static Foo(int i) {}
}
```
https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABATAAgA0BYAKAG8z9r8AHBASwDcJ59sBGANnwDEB7fgAoGAOxj4GASnzkAvmTlA===

**Expected Behavior**:
```
error CS1520: Method must have a return type
```

**Actual Behavior**:
```
error CS0515: 'X.X(int)': access modifiers are not allowed on static constructors
error CS1520: Method must have a return type
error CS0132: 'X.X(int)': a static constructor must be parameterless
```

In this case `Foo` is unlikely to be a failed attempt at a static constructor because (1) it has the wrong name (2) it has parameters (3) it has an access modifier. It is vastly more likely I just forgot the return type on a regular method.

If the method name does not match the type name, the compiler should not give diagnostics for static constructors.

**Comment by CyrusNajmabadi** [comment]
> If the method name does not match the type name, the compiler should not give diagnostics for static constructors.

The main problem there is that that means that parsing is *context sensitive*.  i.e. the parser needs to know something about the container when parsing the members.  This is challenging for several reasons:

1. it makes incremental parsing more difficult.  The parser needs to ensure that these sorts of implicit relationships are understood and mapped over when doing the reparse.
2. it makes it hard to have an API to do things like "parse a member".  You can't just have `ParseMember(text)`, because the parser needs more data to feed in to supply this information.

One potential option here is to change these from parser errors to instead be semantic errors that are checked once we get past the parsing state.  But that may be too much risk/churn for the benefit.

Tagging @gafter as both he and i are pretty interested in this area and in keeping the parser context-free (even if it does lead to this sort of thing at times).

**Comment by gafter** [comment]
This issue appears to be complaining about the diagnostics, not the parse tree.  I agree that we could produce fewer and more focused diagnostics.  However, for reasons that @CyrusNajmabadi discusses, we would continue to parse it as a static constructor.
